### PR TITLE
Add CBMC proof for vSocketClose

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR adds CBMC proof for vSocketClose.

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "FreeRTOS_Sockets.h"
+
+#include "freertos_api.c"
+#include "memory_assignments.c"
+
+void vEventGroupDelete( EventGroupHandle_t xEventGroup )
+{
+    __CPROVER_assert( xEventGroup );
+
+    /**/
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:23]: Delete?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+/* The memory safety of vTCPWindowDestroy has already been proved in
+ * proofs/TCPWin/vTCPWindowDestroy. */
+void vTCPWindowDestroy( TCPWindow_t const * xWindow )
+{
+    /* Do nothing. */
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:30]: Check xWindow is not null?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+void harness()
+{
+    size_t xRequestedSizeBytes;
+    /* Request a random number of bytes keeping in mind the maximum bound of CBMC. */
+    __CPROVER_assume( ( xRequestedSizeBytes >=0 ) && ( xRequestedSizeBytes < ( CBMC_MAX_OBJECT_SIZE - ipBUFFER_PADDING ) ) );
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:37]: Fine, but size_t is always positive.

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    struct xLIST_ITEM temp1;
+
+    List_t temp2;
+    vListInitialise( &temp2 );
+
+    pxSocket->u.xUDP.xWaitingPacketsList.xListEnd.pxNext = &temp1;
+    pxSocket->u.xUDP.xWaitingPacketsList.xListEnd.pxNext->pvOwner = temp;
+    ( ( NetworkBufferDescriptor_t * ) ( pxSocket->u.xUDP.xWaitingPacketsList.xListEnd.pxNext->pvOwner ) )->xBufferListItem.pvContainer = NULL;
+*/
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:61]: Delete commented out code?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    __CPROVER_assume( pxSocket != FREERTOS_INVALID_SOCKET );
+
+    pxSocket->pxUserWakeCallback = safeMalloc( sizeof( SocketWakeupCallback_t ) );
+
+    /* Get a network buffer descriptor with requested bytes. See the constraints
+     * on the number of requested bytes above. */
+    pxSocket->u.xTCP.pxAckMessage = pxGetNetworkBufferWithDescriptor( xRequestedSizeBytes, 0 );
+
+    pxSocket->u.xUDP.xWaitingPacketsList.uxNumberOfItems = 0;
+/*    NetworkBufferDescriptor_t *temp = pxGetNetworkBufferWithDescriptor( 200, 0 );
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:51]: Delete commented out code?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    FreeRTOS_Socket_t * pxSocket = ensure_FreeRTOS_Socket_t_is_allocated();
+    /* Assume socket to not be NULL or invalid. */
+    __CPROVER_assume( pxSocket != NULL );
+    __CPROVER_assume( pxSocket != FREERTOS_INVALID_SOCKET );
+
+    pxSocket->pxUserWakeCallback = safeMalloc( sizeof( SocketWakeupCallback_t ) );
+
+    /* Get a network buffer descriptor with requested bytes. See the constraints
+     * on the number of requested bytes above. */
+    pxSocket->u.xTCP.pxAckMessage = pxGetNetworkBufferWithDescriptor( xRequestedSizeBytes, 0 );
```
Comment (markrtuttle) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:48]: Always 0 for second argument?

**Review by markrtuttle — APPROVED** [review]
Left some comments about minor clean up, but otherwise looks good!

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        vListInitialise( &( pxSocket->u.xUDP.xWaitingPacketsList ) );
+        ListItem_t xWaitingPacketListItem;
+        NetworkBufferDescriptor_t NetworkBuffer;
+
+        if( nondet_bool() )
+        {
+            vListInitialiseItem( &xWaitingPacketListItem );
+            listSET_LIST_ITEM_OWNER( &xWaitingPacketListItem, ( void * ) &NetworkBuffer );
+            vListInsertEnd( &( pxSocket->u.xUDP.xWaitingPacketsList ), &xWaitingPacketListItem );
```
Comment (htibosch) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:79]: HT The item 'xWaitingPacketListItem' has be declared locally in a compound block.
You are adding it to a list that remains to exist outside the block.
Is that what you intend to do?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+            vListInitialiseItem( &( NetworkBuffer.xBufferListItem ) );
+
+            /* Below 2 statements to be checked. */
+            NetworkBuffer.xBufferListItem.pxContainer = &( pxSocket->u.xUDP.xWaitingPacketsList );
+            vListInsertEnd( &( pxSocket->u.xUDP.xWaitingPacketsList ), &( NetworkBuffer.xBufferListItem ) );
```
Comment (htibosch) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:85]: The same for NetworkBuffer: it is declared on the current stack, and it should not be accessed later on in vSocketClose().

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+            pxSocket->u.xUDP.xWaitingPacketsList.uxNumberOfItems = 0;
```
Comment (htibosch) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:87]: I guess that I am looking at CBMB magic now.  Why do you clear `uxNumberOfItems` here?

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        else
+        {
+            pxSocket->u.xUDP.xWaitingPacketsList.uxNumberOfItems = 0;
+        }
+    }
+
+    /* Initialise the item to be used later on in the proof.  */
+    vListInitialiseItem( &( pxSocket->xBoundSocketListItem ) );
```
Comment (htibosch) [inline test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:96]: Here above you have set `xBoundSocketListItem`, now you overwrite these values deliberately?

**Review by htibosch — COMMENTED** [review]
Aniruddha, thanks for these changes.
Maybe my comments only show how little I understand of CBMC yet.
Thanks, Hein

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c (hunk 9 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:87]: This was only for testing... I was trying to figure out why is this proof not running...

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c (hunk 8 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:85]: Good Catch! Thanks Hein!
This is not what I intended to do. I will correct that.

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c (hunk 10 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:96]: Again good catch. Thank you.

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c (hunk 7 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:79]: Good catch! Thanks Hein!

File: test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c (hunk 2 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/Socket/vSocketClose/vSocketClose_harness.c:26]: Done, thanks.

**Comment by AniruddhaKanhere** [comment]
Closing this PR and recreating this in the "main" branch.

**Pull request closed without merging**